lib/fluent/plugin/filter_analyze_config.rb

```diff
@@ -0,0 +1,275 @@
+      # Override instance_eval so that LiteralParser does not actually
+      # evaluate the embedded Ruby, but instead just returns the
+      # source string.  See
+      # https://github.com/fluent/fluentd/blob/master/lib/fluent/config/literal_parser.rb
+      def eval_context.instance_eval(code)
+        code
+      end
+      Fluent::Config::V1Parser.parse(data, fname, basepath, eval_context)
+    end
+
+    # Returns a name for identifying plugins we ship by default.
+    def default_plugin_name(e)
+      case e['@type']
+      when 'syslog'
+        "#{e.name}/syslog/#{e['protocol_type']}"
+      when 'tail'
+        "#{e.name}/tail/#{File.basename(e['pos_file'], '.pos')}"
+      else
+        "#{e.name}/#{e['@type']}"
+      end
+    end
+
+    # Returns a name for identifying plugins not in our default
+    # config.  This should not contain arbitrary user-supplied data.
+    def custom_plugin_name(e)
+      if KNOWN_PLUGINS.key?(e.name) &&
+         KNOWN_PLUGINS[e.name].include?(e['@type'])
+        "#{e.name}/#{e['@type']}"
+      else
+        "#{e.name}"
+      end
+    end
+
+    def embedded_ruby?(e)
+      (e.arg.include?('#{') ||
+       e.any? { |_, v| v.include?('#{') } ||
+       e.elements.any? { |ee| embedded_ruby?(ee) })
+    end
+
+    def configure(conf)
+      super
+      if File.file?(@google_fluentd_config_path) &&
+         File.file?(@google_fluentd_baseline_config_path)
+        @log.info(
+          'google-fluentd configuration file found at' \
+          " #{@google_fluentd_config_path}. " \
+          'google-fluentd baseline configuration file found at' \
+          " #{@google_fluentd_baseline_config_path}. " \
+          'google-fluentd Analyzing configuration.')
+
+        # TODO: Add OpenCensus support.
+        registry = Monitoring::MonitoringRegistryFactory.create(
+          Monitoring::PrometheusMonitoringRegistry.name, nil, nil, nil)
+
+        plugin_usage = registry.counter(
+          :stackdriver_enabled_plugins,
+          [:plugin_name, :is_default_plugin, :has_customized_config],
+          'Enabled plugins')
+        config_usage = registry.counter(
+          :stackdriver_config_usage,
+          [:param, :is_present, :has_default_value],
+          'Parameter usage for Google Cloud plugin')
+
+        config = parse_config(@google_fluentd_config_path)
+        baseline_config = parse_config(@google_fluentd_baseline_config_path)
+
+        # Create hash of all baseline elements by their plugin names.
+        baseline_elements = Hash[baseline_config.elements.collect do |e|
+                                   [default_plugin_name(e), e]
+                                 end]
+        baseline_google_element = baseline_config.elements.find do |e|
+          e['@type'] == 'google_cloud'
+        end
+
+        # Look at each top-level config element and see whether it
+        # matches the baseline value.
+        #
+        # TODO: Does this need to examine nested config elements?
```

Comment:
I don't think we need. All the plugin specification should be at the root level.